src/Components/WebAssembly/Server/src/ComponentsWebAssemblyEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Mime;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.StaticFiles;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// Extensions for mapping Blazor WebAssembly applications.
+    /// </summary>
+    public static class ComponentsWebAssemblyEndpointRouteBuilderExtensions
+    {
+        /// <summary>
+        /// Maps a Blazor webassembly application to the <paramref name="pathPrefix"/>.
+        /// </summary>
+        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/>.</param>
+        /// <param name="pathPrefix">The <see cref="PathString"/> that indicates the prefix for the Blazor application.</param>
+        /// <returns>The <see cref="IEndpointConventionBuilder"/></returns>
+        public static IEndpointConventionBuilder MapBlazorWebAssemblyApplication(this IEndpointRouteBuilder endpoints, PathString pathPrefix)
+        {
+            if (endpoints is null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            var webHostEnvironment = endpoints.ServiceProvider.GetRequiredService<IWebHostEnvironment>();
+
+            var options = CreateStaticFilesOptions(webHostEnvironment.WebRootFileProvider);
+            var appBuilder = endpoints.CreateApplicationBuilder();
+
+            appBuilder.Use(async (ctx, next) =>
+            {
+                var endpoint = ctx.GetEndpoint();
+                try
+                {
+                    // Set the endpoint to null so that static files doesn't discard the path.
+                    ctx.SetEndpoint(null);
+
+                    if (ctx.Request.Path.StartsWithSegments(pathPrefix, out var rest) &&
+                        rest.StartsWithSegments("/_framework"))
+                    {
+                        // At this point we mapped something from the _/framework
+                        ctx.Response.Headers.Append(HeaderNames.CacheControl, "no-cache");
+                    }
+
+                    // This will invoke the static files middleware plugged-in below.
+                    await next();
+
+                }
+                finally
+                {
+                    ctx.SetEndpoint(endpoint);
+                }
+            });
+
+            appBuilder.UseStaticFiles(options);
```

Comment:
Just to check, what are the possible scenarios where this might conflict unexpectedly with other things a developer might be doing? Are they safe to have their own separate `UseStaticFiles` calls with different options elsewhere?

Not saying it's the wrong design - just want to get my understanding right.

Reply:
> Just to check, what are the possible scenarios where this might conflict unexpectedly with other things a developer might be doing? Are they safe to have their own separate `UseStaticFiles` calls with different options elsewhere?
> 
> Not saying it's the wrong design - just want to get my understanding right.

Yes, you can have your own static files middleware and except for the case they are being mapped into the same subpath (or to "/") this should not interfere.

In the future, I want to flow the paths into the host so that it can map exactly the files the blazor app is meant to serve and nothing else. At that point there's no room for ever conflicting with the static files middleware.

Reply:
> except for the case they are being mapped into the same subpath (or to "/") this should not interfere.

In the default "hosted" template case, if someone puts `app.UseStaticFiles()` into their server project because they also want to serve static files from there (independently of their Blazor application), is this likely to result in any conflicts or unexpected behavior?

As long as it's not hugely disruptive it's probably OK for preview2, but just checking we know what to expect.

> In the future, I want to flow the paths into the host so that it can map exactly the files the blazor app is meant to serve and nothing else.

Totally makes sense as a future enhancement.